packages/fuel-indexer-graphql/src/dynamic.rs

```diff
@@ -0,0 +1,457 @@
+    // introspection queries and then pass any non-introspection queries to our
+    // custom query resolver.
+    match dynamic_request.operation_name.as_deref() {
+        Some("IntrospectionQuery") | Some("introspectionquery") => {
+            let introspection_results = dynamic_schema.execute(dynamic_request).await;
+            let data = match introspection_results.data.into_json() {
+                Ok(json) => json,
+                Err(_) => {
+                    return Err(GraphqlError::DynamicQueryError(
+                        "Could not parse introspection response".to_string(),
+                    ))
+                }
+            };
+
+            Ok(serde_json::json!({ "data": data }))
+        }
+        Some(_) | None => {
+            let query =
+                GraphqlQueryBuilder::new(&schema, user_query.as_str())?.build()?;
+
+            let queries = query.as_sql(&schema, pool.database_type())?.join(";\n");
+            let mut conn = match pool.acquire().await {
+                Ok(c) => c,
+                Err(e) => return Err(GraphqlError::DynamicQueryError(e.to_string())),
+            };
+
+            match queries::run_query(&mut conn, queries).await {
+                Ok(val) => Ok(serde_json::json!({ "data": val })),
+                Err(e) => Err(GraphqlError::DynamicQueryError(e.to_string())),
+            }
+        }
+    }
+}
+
+/// Build a dynamic schema. This allows for introspection, which allows for extensive
+/// auto-documentation and code suggestions.
+pub async fn build_dynamic_schema(schema: Schema) -> Result<DynamicSchema, GraphqlError> {
+    let mut schema_builder: DynamicSchemaBuilder =
+        DynamicSchema::build("QueryRoot", None, None)
+            .introspection_only()
+            .register(Scalar::new("Address"))
+            .register(Scalar::new("AssetId"))
+            .register(Scalar::new("Bytes4"))
+            .register(Scalar::new("Bytes8"))
+            .register(Scalar::new("Bytes32"))
+            .register(Scalar::new("Bytes64"))
+            .register(Scalar::new("Int4"))
+            .register(Scalar::new("Int8"))
+            .register(Scalar::new("Int16"))
+            .register(Scalar::new("UInt4"))
+            .register(Scalar::new("UInt8"))
+            .register(Scalar::new("UInt16"))
+            .register(Scalar::new("Timestamp"))
+            .register(Scalar::new("ContractId"))
+            .register(Scalar::new("Salt"))
+            .register(Scalar::new("Json"))
+            .register(Scalar::new("MessageId"))
+            .register(Scalar::new("Charfield"))
+            .register(Scalar::new("Identity"))
+            .register(Scalar::new("Blob"));
+
+    let mut input_objects = Vec::new();
+    let mut filter_object_list = Vec::new();
+    let mut filter_tracker = HashMap::new();
+    let mut sort_object_list = Vec::new();
+    let mut sorter_tracker = HashMap::new();
+
+    let mut query_root = Object::new("QueryRoot");
```

Comment:
Can you mention why we have this `QueryRoot` again? If it's for legacy reasons let's make that clear

Reply:
Yep! Its **only** function there is to allow for introspection and autocomplete at the base query object. For example, without the QueryRoot in the dynamic schema:

query {
  <-- you won't get autocomplete here for anything
}

query {
  block {
    <-- but you do get autocompletions for fields on block 
  }
}

Without that holding object, you don’t get any generated docs and all that nice playground stuff for the entity types. For what it's worth, I did try removing it, but it degrades the experience considerably.

You can see its usage in the example from `async-graphql` here: https://github.com/async-graphql/examples/blob/7eee23ad4e005ac180df8111e6940cebb7c1f52f/models/dynamic-starwars/src/model.rs#L173-L271.

Keep in mind that the `QueryRoot` is not made in the user's indexer schema or even required to be considered by the user at all. This is literally just a requirement by `async-graphql` to be able to serve information about fields when using a dynamic schema, and only matters when an introspection query is sent to the API.